### PR TITLE
Always update LS state through Project

### PR DESCRIPTION
**Pull request by sheetalkamat**
There have been multiple reports of issues with hitting assert at https://github.com/microsoft/TypeScript/blob/80ab111685e3e0c10d8d13cdd18d8dffa8328895/src/server/project.ts#L1518 because the program update before that didnt pass through Project's update graph and hence LS and Project go out of sync, Recent report at https://github.com/microsoft/TypeScript/pull/55716#issuecomment-1752163519

It has always been hard to reproduce this without plugin and race conditions etc. So for testing i wrote simple plugin test that fails without this fix.

### CODE REVIEW

File: src/testRunner/unittests/tsserver/plugins.ts [hunk 1]
```diff
@@ -101,6 +103,38 @@ describe("unittests:: tsserver:: plugins:: loading", () => {
...
+                        { name: pluginName },
+                    ],
+                },
+            }),
+        };
+
+        const { session, host } = createHostWithPlugin([aTs, tsconfig, libFile]);
+
+        openFilesForSession([aTs], session);
+        host.writeFile("/user/username/projects/project/b.ts", "const y = 10;"); // Missing file is writting so scheduling the update
```
Comment (DanielRosenwasser) [inline src/testRunner/unittests/tsserver/plugins.ts:126]:
```suggestion
        // Write the missing file (referenced by 'a.ts') to schedule an update.
        host.writeFile("/user/username/projects/project/b.ts", "const y = 10;");
```

File: src/testRunner/unittests/tsserver/plugins.ts [hunk 2]
```diff
@@ -101,6 +103,38 @@ describe("unittests:: tsserver:: plugins:: loading", () => {
...
+
+        session.executeCommandSeq({ // This should update LS with new program
```
Comment (DanielRosenwasser) [inline src/testRunner/unittests/tsserver/plugins.ts:128]:
```suggestion
        // This should update the language service with a new program.
        session.executeCommandSeq({
```

**Comment by jakebailey** [comment]
@typescript-bot pack this

@johnsoncodehk does this fix the crash you noted in https://github.com/microsoft/TypeScript/pull/55716#issuecomment-1752163519 ?

File: src/testRunner/unittests/tsserver/plugins.ts [hunk 3]
```diff
@@ -101,6 +103,38 @@ describe("unittests:: tsserver:: plugins:: loading", () => {
...
+            command: testProtocolCommand,
+            arguments: testProtocolCommandRequest,
+        });
+
+        host.runQueuedTimeoutCallbacks(); // This results in program update
```
Comment (DanielRosenwasser) [inline src/testRunner/unittests/tsserver/plugins.ts:133]:
```suggestion
        // This results in a program update.
        host.runQueuedTimeoutCallbacks();
```

File: tests/baselines/reference/tsserver/plugins/when-plugins-use-LS-to-get-program-and-update-is-pending.js [hunk 4]
```diff
@@ -0,0 +1,318 @@
+currentDirectory:: / useCaseSensitiveFileNames: false
+Info seq  [hh:mm:ss:mss] Provided types map file "/typesMap.json" doesn't exist
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/tsserver/plugins/when-plugins-use-LS-to-get-program-and-update-is-pending.js:2]: Is this a problem?

Comment (sheetalkamat) [reply tests/baselines/reference/tsserver/plugins/when-plugins-use-LS-to-get-program-and-update-is-pending.js:2]: No.. Its used only in some of the TI test cases and there that file is on the host

**Comment by johnsoncodehk** [comment]
> @johnsoncodehk does this fix the crash you noted in [#55716 (comment)](https://github.com/microsoft/TypeScript/pull/55716#issuecomment-1752163519) ?

@jakebailey I can confirm this fix works, thanks!

**Review by jakebailey — APPROVED** [review]

**Pull request merged**